### PR TITLE
Record iOS text in SR

### PR DESCRIPTION
**Pull request by louiszawadzki**
### What does this PR do?

Enable recording of texts in SR.

### Additional Notes

For testing I decided to include a bit of mocking of RN to be able to cover the node recorder. 
However I find these mocks to be a bit too tied to the actual implementation, so we should be covered later with more E2E tests and we can potentially get rid of these ones.

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)
- [ ] If this PR is auto-generated, please make sure also to manually update the code related to the change

### CODE REVIEW

File: example/ios/PodfileForSRTests
Comment (louiszawadzki) [inline example/ios/PodfileForSRTests:24]: When running iOS SR Tests on Bitrise, I switch the Podfile with this file so we can run the tests.

`use_modular_headers!` is needed for SR tests as `React-RCTText` does not declare that it declares modules.

File: example/ios/PodfileForSRTests
Comment (louiszawadzki) [inline example/ios/PodfileForSRTests]: This is to temporarily enable to run the tests with the latest iOS SDK version

File: packages/react-native-session-replay/DatadogSDKReactNativeSessionReplay.podspec
Comment (louiszawadzki) [inline packages/react-native-session-replay/DatadogSDKReactNativeSessionReplay.podspec:25]: I did not really understand why this was only needed for tests. If I'm not adding this the tests fail with a `Undefined symbols for architecture arm64: "_OBJC_CLASS_$_RCTTextView"` error.
However running an app works fine. I assumed this was some RN black magic, but I'd be happy to get a second opinion on this.

As `React-RCTText` does not declare modules, pod install will fail unless `use_modular_headers!` is used in the podfile, so I only add the dependency to the test spec so our users are not impacted.

File: packages/react-native-session-replay/ios/Sources/RCTTextViewRecorder.swift
Comment (louiszawadzki) [inline packages/react-native-session-replay/ios/Sources/RCTTextViewRecorder.swift:60]: See https://datadoghq.atlassian.net/l/cp/FHhsjKoo for more context on other potential solutions with perf benchmarks

File: packages/react-native-session-replay/ios/Tests/RCTTextViewRecorderTests.swift
Comment (louiszawadzki) [inline packages/react-native-session-replay/ios/Tests/RCTTextViewRecorderTests.swift]: this feels a bit hacky, if you see a better way to do that I'm happy to know about it

**Comment by louiszawadzki** [comment]
Example of texts replay with several cases:

Simulator:
image

Replay (obfuscated):
image

Replay (not obfuscated):
image

File: packages/react-native-session-replay/ios/Sources/RCTTextViewRecorder.swift
Comment (maciejburda) [inline packages/react-native-session-replay/ios/Sources/RCTTextViewRecorder.swift:62]: Maybe it's cleaner to just extract the text from this view and change subtreeStrategy to record?

File: packages/react-native-session-replay/ios/Sources/RCTTextViewRecorder.swift
Comment (maciejburda) [inline packages/react-native-session-replay/ios/Sources/RCTTextViewRecorder.swift:72]: It actually goes to the player level.

Is font family output by RN web-safe? If no - there's work to be done on RN side as well.

File: packages/react-native-session-replay/ios/Sources/RCTTextViewRecorder.swift
Comment (maciejburda) [inline packages/react-native-session-replay/ios/Sources/RCTTextViewRecorder.swift]: Personal preference, but I'd say it's fine to drop these `///` they don't add much context.

File: packages/react-native-session-replay/ios/Tests/RCTTextViewRecorderTests.swift
Comment (maciejburda) [reply packages/react-native-session-replay/ios/Tests/RCTTextViewRecorderTests.swift]: You can shorten it this way, but unfortunately, that's how deal with associated values in enums:

```swift
extension SRWireframe {
    public func getAsTextWireframe() -> SRTextWireframe? {
        if case .textWireframe(let value) = self {
            return value
        }
        return nil
    }
}
```

File: packages/react-native-session-replay/ios/Tests/RCTTextViewRecorderTests.swift
Comment (maciejburda) [inline packages/react-native-session-replay/ios/Tests/RCTTextViewRecorderTests.swift:13]: exploring snapshot tests for RN would be super useful

File: packages/react-native-session-replay/ios/Tests/RCTTextViewRecorderTests.swift
Comment (maciejburda) [inline packages/react-native-session-replay/ios/Tests/RCTTextViewRecorderTests.swift]:
```suggestion
        // The shadow view must be initialized with a bridge so that we can insert React Subviews into it.
```

**Review by maciejburda — COMMENTED** [review]
Gave it a first pass.

Looking good. I will take a closer look in the morning.

File: packages/react-native-session-replay/ios/Sources/RCTTextViewRecorder.swift
Comment (louiszawadzki) [reply packages/react-native-session-replay/ios/Sources/RCTTextViewRecorder.swift:62]: I think this is not possible here because of the way RN works.
If you take this example:
```jsx
<Text>this is a <Text>nested text</Text></Text>
```

You would have 1 `RCTTextView` (`UIView`) for all the nested text, with 1 corresponding `RCTTextShadowView` (not a `UIView`) that contains the subtree of nested shadow views like so:

```
RCTTextShadowView
| RCTRawTextShadowView (text: "this is a ")
| RCTVirtualTextShadowView
  | RCTRawTextShadowView (text: "nested text")
```

If we were to change subtreeStrategy to record, we would not find any nested view since we only traverse `UIView` tree, and not this shadow view tree.

Comment (louiszawadzki) [reply packages/react-native-session-replay/ios/Sources/RCTTextViewRecorder.swift:62]: I believe RN uses an attributedString (which isn't public) on the UIView to achieve the rendering.

File: packages/react-native-session-replay/ios/Sources/RCTTextViewRecorder.swift
Comment (louiszawadzki) [reply packages/react-native-session-replay/ios/Sources/RCTTextViewRecorder.swift:72]: That's a good question, I've not looked into it yet. Usually you add font files, named whatever you want, with 1 file per variant (weight + style).
So I'd say it's the same as what is done for iOS, not web safe

File: packages/react-native-session-replay/ios/Tests/RCTTextViewRecorderTests.swift
Comment (louiszawadzki) [reply packages/react-native-session-replay/ios/Tests/RCTTextViewRecorderTests.swift:13]: Yes, I want to explore this in 2 weeks. 
My other plan is to do something similar to what Xavier is doing on Android and build a DatadogCore that just writes messages to memory and that you can query later for assertions, to be used in E2E tests instead of monitors.

File: packages/react-native-session-replay/ios/Sources/RCTTextViewRecorder.swift
Comment (maciejburda) [reply packages/react-native-session-replay/ios/Sources/RCTTextViewRecorder.swift:62]: Maybe all we need is `RCTRawTextShadowViewRecorder` ?

File: packages/react-native-session-replay/ios/Sources/RCTTextViewRecorder.swift
Comment (maciejburda) [reply packages/react-native-session-replay/ios/Sources/RCTTextViewRecorder.swift:72]: Alright!

Would be great to add a ticket for RN that's blocked by the iOS one

File: packages/react-native-session-replay/ios/Sources/RCTTextViewRecorder.swift
Comment (louiszawadzki) [reply packages/react-native-session-replay/ios/Sources/RCTTextViewRecorder.swift:62]: I can try to catch it this way, but I'm afraid we won't see it as I believe it is not in the UI View tree that is traversed

Comment (louiszawadzki) [reply packages/react-native-session-replay/ios/Sources/RCTTextViewRecorder.swift:62]: Indeed it won't ever work:
image

Comment (maciejburda) [reply packages/react-native-session-replay/ios/Sources/RCTTextViewRecorder.swift:62]: I see. It almost asks for separate set of recorders that would work on those RN views in that case.

I mean - I am happy with the current solution. Wondering if we can do anything better in the future when we scale up number of the views.

**Review by maciejburda — APPROVED** [review]

**Review by maciejburda — APPROVED** [review]
Just gave it another look, and tested locally.

LGTM!

**Pull request merged**
